runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -374,6 +374,62 @@ function M.get(bufnr, client_id)
   return diagnostic_cache[bufnr][client_id] or {}
 end
 
+--- Get the diagnostics by cursor
+---
+---@param bufnr number The buffer number
+---@param cursor table The cursor
+---@param opts table|nil Configuration keys
+---         - severity: (DiagnosticSeverity, default nil)
+---             - Only return diagnostics with this severity. Overrides severity_limit
+---         - severity_limit: (DiagnosticSeverity, default nil)
+---             - Limit severity of diagnostics found. E.g. "Warning" means { "Error", "Warning" } will be valid.
+---@param client_id number the client id
+---@return table Table with map of line number to list of diagnostics.
+--               Structured: { [1] = {...}, [5] = {.... } }
+function M.get_cursor_diagnostics(bufnr, cursor, opts, client_id)
```

Comment:
I would refactor this to take a position instead of "cursor" and rename the M.get_position_diagnostics, then default as you have currently to using the current win position.